File: backend/models/Room.js
const mongoose = require('mongoose');

/**
 * Room Schema for Group Chat
 * 
 * Security Design:
 * - passwordHash: bcrypt hash, never store plain-text password
 * - sessionKeyVersion: Increments when key is rotated (e.g., after kicking member)
 * - Each member stores their own encrypted copy of the session key
 * 
 * Key Distribution Flow:
 * 1. Admin creates group -> generates random SessionKey
 * 2. Admin encrypts SessionKey with each member's public key
 * 3. Server stores encrypted copies (one per member)
 * 4. When user joins, server sends their encrypted copy
 * 5. User decrypts with their private key -> Gets SessionKey
 */
const roomSchema = new mongoose.Schema({
  // Room metadata
  name: {
    type: String,
    required: true,
    trim: true,
    maxlength: 100
  },
  
  type: {
    type: String,
    enum: ['group', 'channel'], // 'group' for group chat, 'channel' for broadcast
    default: 'group',
    required: true
  },
  
  avatar: {
    type: String, // URL to group avatar
    default: null
  },
  
  description: {
    type: String,
    maxlength: 500,
    default: ''
  },
  
  // Owner/Admin management
  ownerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  
  // Password protection
  passwordHash: {
    type: String, // bcrypt hash of room password
    default: null // null = public room (no password required)
  },
  
  isPasswordProtected: {
    type: Boolean,
    default: false
  },
  
  // Members
  members: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  
  memberCount: {
    type: Number,
    default: 0
  },
  
  // E2EE Key Management
  sessionKeyVersion: {
    type: Number,
    default: 1,
    // Increments every time we rotate the key (e.g., after kicking a member)
    // Clients can check if they have the latest key version
  },
  
  // Settings
  settings: {
    allowMembersToInvite: {
      type: Boolean,
      default: true
    },
    allowMembersToAddMembers: {
      type: Boolean,
      default: false // Only admin can add by default
    },
    maxMembers: {
      type: Number,
      default: 500
    }
  },
  
  // Timestamps
  createdAt: {
    type: Date,
    default: Date.now
  },
  
  updatedAt: {
    type: Date,
    default: Date.now
  },
  
  lastMessageAt: {
    type: Date,
    default: Date.now
  }
});

// Indexes for performance
roomSchema.index({ ownerId: 1 });
roomSchema.index({ members: 1 });
roomSchema.index({ createdAt: -1 });
roomSchema.index({ lastMessageAt: -1 });

// Update timestamp before saving
roomSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

// Methods
roomSchema.methods = {
  // Check if user is member
  isMember(userId) {
    return this.members.some(memberId => memberId.toString() === userId.toString());
  },
  
  // Check if user is owner
  isOwner(userId) {
    return this.ownerId.toString() === userId.toString();
  },
  
  // Add member to room
  async addMember(userId) {
    if (!this.isMember(userId)) {
      this.members.push(userId);
      this.memberCount = this.members.length;
      await this.save();
    }
  },
  
  // Remove member from room
  async removeMember(userId) {
    const index = this.members.findIndex(memberId => memberId.toString() === userId.toString());
    if (index > -1) {
      this.members.splice(index, 1);
      this.memberCount = this.members.length;
      await this.save();
    }
  },
  
  // Rotate key version (after kicking member)
  async rotateSessionKey() {
    this.sessionKeyVersion += 1;
    await this.save();
  }
};

// Statics
roomSchema.statics = {
  // Get rooms where user is member
  async getUserRooms(userId) {
    return this.find({ members: userId })
      .populate('ownerId', 'username email avatar')
      .sort({ lastMessageAt: -1 });
  },
  
  // Get public rooms (no password)
  async getPublicRooms(limit = 20, skip = 0) {
    return this.find({ 
      isPasswordProtected: false,
      isPrivate: false  // Only show truly public groups
    })
      .populate('ownerId', 'username email avatar')
      .limit(limit)
      .skip(skip)
      .sort({ memberCount: -1, createdAt: -1 });
  }
};

module.exports = mongoose.model('Room', roomSchema);

File: flutter/lib/models/room.dart
/**
 * Room Model - Represents a Group Chat Room
 * 
 * Corresponds to backend Room schema
 */
class Room {
  final String id;
  final String name;
  final String? avatar;
  final String? description;
  final String type; // 'group' or 'channel'
  final String ownerId;
  final bool isPasswordProtected;
  final bool isPrivate; // Whether group is private or public
  final int memberCount;
  final int sessionKeyVersion;
  final DateTime createdAt;
  final DateTime? lastMessageAt;
  final RoomSettings? settings;

  Room({
    required this.id,
    required this.name,
    this.avatar,
    this.description,
    required this.type,
    required this.ownerId,
    required this.isPasswordProtected,
    this.isPrivate = false,
    required this.memberCount,
    required this.sessionKeyVersion,
    required this.createdAt,
    this.lastMessageAt,
    this.settings,
  });

  factory Room.fromJson(Map<String, dynamic> json) {
    // Helper to safely get string value
    String getStringValue(dynamic value) {
      if (value == null) return '';
      if (value is String) return value;
      return value.toString();
    }
    
    return Room(
      id: getStringValue(json['id'] ?? json['_id']),
      name: getStringValue(json['name']),
      avatar: json['avatar']?.toString(),
      description: json['description']?.toString(),
      type: getStringValue(json['type'] ?? 'group'),
      ownerId: getStringValue(
        json['ownerId'] ?? json['owner']?['id'] ?? json['owner']?['_id'] ?? json['owner']
      ),
      isPasswordProtected: json['isPasswordProtected'] ?? false,
      isPrivate: json['isPrivate'] ?? false,
      memberCount: (json['memberCount'] ?? 0) as int,
      sessionKeyVersion: (json['sessionKeyVersion'] ?? 1) as int,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'].toString())
          : DateTime.now(),
      lastMessageAt: json['lastMessageAt'] != null
          ? DateTime.parse(json['lastMessageAt'].toString())
          : null,
      settings: json['settings'] != null
          ? RoomSettings.fromJson(json['settings'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'avatar': avatar,
      'description': description,
      'type': type,
      'owner_id': ownerId,
      'is_password_protected': isPasswordProtected ? 1 : 0,
      'member_count': memberCount,
      'session_key_version': sessionKeyVersion,
      'created_at': createdAt.toIso8601String(),
      'last_message_at': lastMessageAt?.toIso8601String(),
    };
  }

  // SQLite database methods
  factory Room.fromDatabase(Map<String, dynamic> row) {
    return Room(
      id: row['id'] as String,
      name: row['name'] as String,
      avatar: row['avatar'] as String?,
      description: row['description'] as String?,
      type: row['type'] as String,
      ownerId: row['owner_id'] as String,
      isPasswordProtected: (row['is_password_protected'] as int) == 1,
      isPrivate: (row['is_private'] as int?) == 1,
      memberCount: row['member_count'] as int,
      sessionKeyVersion: row['session_key_version'] as int,
      createdAt: DateTime.parse(row['created_at'] as String),
      lastMessageAt: row['last_message_at'] != null
          ? DateTime.parse(row['last_message_at'] as String)
          : null,
      settings: null, // Settings loaded separately if needed
    );
  }

  Map<String, dynamic> toDatabase() {
    return {
      'id': id,
      'name': name,
      'avatar': avatar,
      'description': description,
      'type': type,
      'owner_id': ownerId,
      'is_password_protected': isPasswordProtected ? 1 : 0,
      'is_private': isPrivate ? 1 : 0,
      'member_count': memberCount,
      'session_key_version': sessionKeyVersion,
      'created_at': createdAt.toIso8601String(),
      'last_message_at': lastMessageAt?.toIso8601String(),
    };
  }

  Room copyWith({
    String? id,
    String? name,
    String? avatar,
    String? description,
    String? type,
    String? ownerId,
    bool? isPasswordProtected,
    bool? isPrivate,
    int? memberCount,
    int? sessionKeyVersion,
    DateTime? createdAt,
    DateTime? lastMessageAt,
    RoomSettings? settings,
  }) {
    return Room(
      id: id ?? this.id,
      name: name ?? this.name,
      avatar: avatar ?? this.avatar,
      description: description ?? this.description,
      type: type ?? this.type,
      ownerId: ownerId ?? this.ownerId,
      isPasswordProtected: isPasswordProtected ?? this.isPasswordProtected,
      isPrivate: isPrivate ?? this.isPrivate,
      memberCount: memberCount ?? this.memberCount,
      sessionKeyVersion: sessionKeyVersion ?? this.sessionKeyVersion,
      createdAt: createdAt ?? this.createdAt,
      lastMessageAt: lastMessageAt ?? this.lastMessageAt,
      settings: settings ?? this.settings,
    );
  }
}

/**
 * Room Settings Model
 */
class RoomSettings {
  final bool allowMembersToInvite;
  final bool allowMembersToAddMembers;
  final int maxMembers;

  RoomSettings({
    required this.allowMembersToInvite,
    required this.allowMembersToAddMembers,
    required this.maxMembers,
  });

  factory RoomSettings.fromJson(Map<String, dynamic> json) {
    return RoomSettings(
      allowMembersToInvite: json['allowMembersToInvite'] ?? true,
      allowMembersToAddMembers: json['allowMembersToAddMembers'] ?? false,
      maxMembers: json['maxMembers'] ?? 500,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'allowMembersToInvite': allowMembersToInvite,
      'allowMembersToAddMembers': allowMembersToAddMembers,
      'maxMembers': maxMembers,
    };
  }
}
